move/interchain_token_service/sources/versioned/interchain_token_service_v0.move

```diff
@@ -181,6 +181,36 @@ module interchain_token_service::interchain_token_service_v0 {
         coin_management: CoinManagement<T>,
     ): TokenId {
         let token_id = token_id::from_coin_data(&self.chain_name_hash, &coin_info, &coin_management);
+        let mut coin_data = coin_data::new(coin_management, coin_info);
+        let coin_info = coin_data.coin_info_mut();
+
+        // This will no-op if coin_info.metadata is None
+        coin_info.release_metadata();
+
+        self.add_registered_coin(token_id, coin_data);
+
+        token_id
+    }
+
+    public(package) fun register_coin_from_info<T>(
+        self: &mut InterchainTokenService_v0,
+        name: std::string::String,
+        symbol: ascii::String,
+        decimals: u8,
+        coin_management: CoinManagement<T>,
+    ): TokenId {
+        let coin_info = coin_info::from_info<T>(name, symbol, decimals);
+
+        self.register_coin(coin_info, coin_management)
+    }
+
+    public(package) fun register_coin_from_metadata<T>(
+        self: &mut InterchainTokenService_v0,
+        metadata: &CoinMetadata<T>,
+        coin_management: CoinManagement<T>,
+    ): TokenId {
+        let coin_info = coin_info::from_metadata_ref<T>(metadata);
+        let token_id = token_id::from_coin_data(&self.chain_name_hash, &coin_info, &coin_management);
 
         self.add_registered_coin(token_id, coin_data::new(coin_management, coin_info));
```

Comment:
Call register_coin instead of duplicating the logic. Pass a new bool arg to register_coin that indicates whether coin metadata was used or not to create coin info. Use this bool in the token derivation